### PR TITLE
[Rocketmq-281] add check policy for preventing repeat start mq

### PR DESCRIPTION
**Pull request by fuyou001**
To avoid repeat start mq.
Motivation:
 add a filelock check when mq start, if lock success ,start success,otherwise start failed.

Motivation:
brokerStartup add filelock check

### CODE REVIEW

**Comment by zhouxinyu** [comment]
Please submit PR to `develop` branch~

File: broker/src/main/java/org/apache/rocketmq/broker/BrokerStartup.java [hunk 1]
```diff
@@ -257,4 +268,28 @@ public static Options buildCommandlineOptions(final Options options) {
 
         return options;
     }
+
+    private static void checkMQAlreadyStart(String storePathRootDir) throws IOException {
```
Comment (zhouxinyu) [inline broker/src/main/java/org/apache/rocketmq/broker/BrokerStartup.java]: What's the relation between `broker` and `store root path`? I would recommend that check whether `store` is available in `MessageStore` module, if `MessageStore` can't start properly, then exit~

File: broker/src/main/java/org/apache/rocketmq/broker/BrokerStartup.java [hunk 2]
```diff
@@ -257,4 +268,28 @@ public static Options buildCommandlineOptions(final Options options) {
...
+        final RandomAccessFile randomAccessFile = new RandomAccessFile(file, "rw");
+        FileLock lock = randomAccessFile.getChannel().tryLock(0, 1, false);
+        if (lock == null || lock.isShared() || !lock.isValid()) {
+            throw new RuntimeException("Lock failed,MQ already started");
+        }
+
+        randomAccessFile.getChannel().write(ByteBuffer.wrap("lock".getBytes()));
+        randomAccessFile.getChannel().force(true);
+
+        Runtime.getRuntime().addShutdownHook(new Thread() {
```
Comment (zhouxinyu) [inline broker/src/main/java/org/apache/rocketmq/broker/BrokerStartup.java]: Also, release the file lock in `shutdown` of `MessageStore` is recommended.

Comment (lindzh) [reply broker/src/main/java/org/apache/rocketmq/broker/BrokerStartup.java]: +1

**Comment by lindzh** [comment]
After merge develop branch,all checks will pass

File: broker/src/main/java/org/apache/rocketmq/broker/BrokerStartup.java (hunk 1 above)
Comment (fuyou001) [reply broker/src/main/java/org/apache/rocketmq/broker/BrokerStartup.java]: done

File: broker/src/main/java/org/apache/rocketmq/broker/BrokerStartup.java (hunk 2 above)
Comment (fuyou001) [reply broker/src/main/java/org/apache/rocketmq/broker/BrokerStartup.java]: done

**Comment by zhouxinyu** [comment]
LGTM now, please help resolve conflicting files~

**Comment by penghuazhou** [comment]
11

**Pull request closed without merging**